refactor: rewrite command-not-found.sh in rust

figsoda opened the pull request:
closes #222 
closes #126
should also fix #210 once this is finished

the implementation is adapted from `locate` and `command-not-found.sh`

draft because there are a few things I need to do:
- [x] actually replace the command-not-found.sh, including fish support
- [ ] complete `has_flakes` to read user config, also consider keeping the previous `manifest.json` check
- [ ] better error handling, currently some errors are silently ignored
- [ ] make the code a bit more compact, the `Command::new`s feel a bit verbose

figsoda committed:
refactor: rewrite command-not-found.sh in rust

emilazy commented:
Wonderful!

The fish mechanism [is simple](https://fishshell.com/docs/current/cmds/fish_command_not_found.html) but there are some nuances to consider that break the existing script (translated with babelfish); see https://github.com/fish-shell/fish-shell/issues/9806, https://github.com/fish-shell/fish-shell/issues/7902, https://github.com/fish-shell/fish-shell/pull/9517. (tl;dr it's run inside a substitution so you can't check tty through stdout, which might also interfere with the auto-run feature in terms of what stdout they get and how the exit status is handled. I wonder if anyone actually uses auto-run though? comma seems like a less error-prone way of getting essentially the same result...)

figsoda commented:
I just found a typo in the auto run error messages that's existed for a few years now

https://github.com/nix-community/nix-index/blob/73d96c85bc1bfe7c0cbd6269ea42119f65073cf9/command-not-found.sh#L56
https://github.com/nix-community/nix-index/blob/73d96c85bc1bfe7c0cbd6269ea42119f65073cf9/command-not-found.sh#L69

that's probably a sign that people are not really using this feature

emilazy commented:
Yeah, I would be happy to see it go. I think the main adaptation you'll need to make for fish is conditioning the pipe check on stdin or stderr, which unfortunately won't be as reliable (e.g. `bad-command | ...`). Not sure if there's a better check that could be done on the fish side and passed in via environment or something. Maybe it's fine to just check stderr and output there though? Seems okay for a piped command to print to stderr and then fail.

figsoda commented:
tbh I think it makes sense for this to run even if it is being piped into another command, maybe we can just remove the check altogether? @bennofs what do you think?

emilazy commented:
Yeah I dunno exactly what the rationale is if no output is going to stdout anyway. I think checking stderr would be good because you're printing output there that could be misinterpreted if you're piping it into another command with `2>&1` or similar.

(fwiw I think NIX_AUTO_INSTALL is even worse than NIX_AUTO_RUN – permanent imperative changes to system configuration in response to what could be a mere typo...)

bennofs commented:
I agree that `NIX_AUTO_RUN` / `NIX_AUTO_INSTALL` are probably fringe usecases that we can drop from upstream (people who want this behavior can use comma / use their own command-not-found handler).

We should include a message if `NIX_AUTO_RUN` / `NIX_AUTO_INSTALL` is set that explains the alternatives (comma, custom command not found) so users are not surprised that the feature is suddenly no longer available.

figsoda committed:
fix(cnf): escape command

figsoda committed:
refactor: replace command-not-found.sh with `nix-locate --command-not-found`

figsoda committed:
fix(nix-locate): compatibility with rust 1.69

figsoda committed:
fix(cnf): typo

figsoda committed:
fix(cnf): gracefully handle flags

figsoda committed:
flake.lock: Update

Flake lock file updates:

• Updated input 'nixpkgs':
    'github:NixOS/nixpkgs/75a5ebf473cd60148ba9aec0d219f72e5cf52519' (2023-06-11)
  → 'github:NixOS/nixpkgs/e603dc5f061ca1d8a19b3ede6a8cf9c9fcba6cdc' (2023-06-22)

figsoda committed:
Revert "fix(nix-locate): compatibility with rust 1.69"

This reverts commit 2030741bf37aa5bb4ec84cf78bb428dc4d489d50.

figsoda committed:
feat(cnf): reduce stderr locking

dsully commented:
Any progress here? Would love to see this merged. Thanks
